### PR TITLE
Add DB schema enhancement changes

### PR DESCRIPTION
**Pull request by alan-amida**
DB Schema changes based on Dimitri's [DB Enhancement Proposa](https://github.com/department-of-veterans-affairs/abd-vro/wiki/VRO-Database#proposed-redesign)l and Yoom's [feedback](https://github.com/department-of-veterans-affairs/abd-vro/wiki/VRO-Database#feedback-on-proposal)

## Current order of operations (subject to change based on review/feedback):
1. Create `veteran_flash_id` table to stored a veteran's associated `flashId`s
  a. currently only processing one flashId for Agent Orange exposure
2. Create `exam_order` table to store the status of `/v2/examOrderingStatus` requests
3. Update `audit_event` table's `details` column type from VARCHAR to JSONB
4. Update our `claim` table with columned fields stored as JOSNB in `audit_event.details` (no replacement):
  a. add `vbms_id` field
  b. add `off_ramp_reason` field
  c. add `presumptive_flag` field
  d. add `disability_action_type` field
  e. add `in_scope` field
  f. add `submission_source` field
  g. add `submission_date` field
5. Update our `veteran` table to include a `icn_timestamp` field
6. Update our `evidence_summary_document` table to include a `location` field

## In-progress operations:
- Implementing Java Entities and necessary other Java code complete the "from Data hitting our API to our DB" path

I want to make sure the Model Schema changes are good with everyone before we proceed with spending any time writing the associated Java.

## Open Questions:
- `claim.id` is our internal DB primary key for the claim, but `claim.claim_submission_id` is an external number provided by [some external caller of our API]?
- Is it better to think of `claim` table as "current claim state" vs `claim_submission` table as the new "claim audit event history"?
  - This might mean we can continue to store `details` as recorded history in `claim_submission`, but continue to maintain current state in `claim` table. Ultimately must consult with Yoom, but might be better to err on the side of redundancy than otherwise.
  - It feels like the root of this ambiguity is because there may be a misunderstanding on (or conflation with) claim history tracking vs current state
- Do the data types look okay?
- Where do we want any more indexing to happen?
- What else am I missing?

## Some clarifications from Dimitri:
- Leave `audit_event` as-is, change `details` to JSONB, just add fields to `claim` table. We can leave out:
  - ~~requestType: CLAIM_SUBMISSION, AUTOMATED_CLAIM, OTHER~~
  - ~~status: PROCESSED, REJECTED, ERROR (or something like this)~~
  - ~~"Make EvidenceSummary and AssessmentResult direct children of ClaimSubmission since these may be different for each invocation"~~
- `vbmsId` is VA.gov id for the claim -- ambiguous on whether claimsubmissionid is govt VA ID or if it is an ID specific to the MAS API (Ask Rajesh?)
  - Rajesh confirms that `vbmsId` is what we get as the `claimSubmissionId/claimId`
- We get `flashId`s as part of the initial payload, we can grab this very early in the process.
- Look in `service/provider/src/main/java/gov/va/vro/service/provider/mas/service/MasProcessingService.java` to intercept the initial submission data

### CODE REVIEW

**Comment by alan-amida** [comment]
I think I'm satisfied with the current schema migration changes. All but one change are adds, the exception being replacing the old `audit_event`details` type from `VARCHAR` to `JSONB` which requires a drop and a re-add.

This minimizes any DB impact, leaving the rest of the changes to be made in Java code. I am breaking ground on those as we speak, starting from the Entities (and associated child-entities) and working my way out to the Repositories, until eventually any broken tests.

**Comment by alan-amida** [comment]
OK. As far as the `audit_event.details` changes go:

- [ ] Tests pass
- [ ] Linting passes
- [ ] Verified migrations script work
- [ ] the new JSONB column type appears to be working with some simple test-cases for `/v2/examOrderingStatus` and `/v2/automatedClaim`

This ⬆️ was the most complex one since it involved the most type reference updates throughout the code.
- [ ] Next up is probably to make the `order_exam` table functional, simply passing the `collectionId` and `collectionStatus` to the table
- [ ] The rest are simply grabbing data and inserting them into new columns in various tables (including the `veteran_flash_id` table

**Comment by alan-amida** [comment]
@yoomlam this is still a WIP, but we're close to 90% done. I still have some changes to push up.

Here, you can see the direction we've taken. Please review the notes.

I consulted with Dimitri on the audit_event table and after I tried to parse intent from previous discussions, I assumed that you meant you just wanted additional fields to help query data easier, but that it didn't mean we needed to *replace* the `audit_event` table.

As such, we felt it appropriate to keep `audit_table` the same (save for one change), and the rest are straight up new field additions to existing tables (plus a `veteran_flash_id` association table).

I'm hoping that you'll look at the schema changes, and maybe take a look at how we're assigning the data to the various objects.fields as a sanity-check.

**Comment by yoomlam** [comment]
Quick response without reviewing the code:
I was under the impression that the [Proposed Redesign](https://github.com/department-of-veterans-affairs/abd-vro/wiki/VRO-Database#proposed-redesign) would be implemented, where:
> Renamed AuditEvent to ClaimSubmission

so there would be no mention of `AuditEvent`.

Dimitri [updated that wiki section](https://dsva.slack.com/archives/C01Q7979Z7D/p1674229300508929?thread_ts=1674144499.310049&cid=C01Q7979Z7D), then I presented my feedback based on those updates.

To help me understand the intent of this PR and follow-on PRs, of the items listed in the "Proposed Redesign" section, what is being implemented and what is not being implemented?

**Comment by alan-amida** [comment]
@yoomlam 
> Quick response without reviewing the code: I was under the impression that the [Proposed Redesign](https://github.com/department-of-veterans-affairs/abd-vro/wiki/VRO-Database#proposed-redesign) would be implemented, where:
> 
> > Renamed AuditEvent to ClaimSubmission
> 
> so there would be no mention of `AuditEvent`.
> 
> Dimitri [updated that wiki section](https://dsva.slack.com/archives/C01Q7979Z7D/p1674229300508929?thread_ts=1674144499.310049&cid=C01Q7979Z7D), then I presented my feedback based on those updates.
> 
> To help me understand the intent of this PR and follow-on PRs, of the items listed in the "Proposed Redesign" section, what is being implemented and what is not being implemented?

Again, review all the notes and comments in this PR. Much of the initial proposal concering a `claim_submission` was tossed out. The Wiki is outdated. This is what we're doing now.

Basically I believe there was a fundamental misunderstanding between the both of you concerning intent regarding audit_event and claim_submission -- after discussing with Dimitri, he agrees.

He thought you meant to replace AuditEvent with the Claim table (replacing fields and putting them in Claim), hence his calling it ClaimSubmission table. He disagreed with "putting those fields in Claim table" under the assumption they'd be moved. After trying to parse intent, I hypothesized that you simply meant to have them tracked as current state, with no mention of removing/replacing the original fields. He later agreed.

> I consulted with Dimitri on the audit_event table and after I tried to parse intent from previous discussions, I assumed that you meant you just wanted additional fields to help query data easier, but that it didn't mean we needed to replace the audit_event table.
> As such, we felt it appropriate to keep audit_table the same (save for one change), and the rest are straight up new field additions to existing tables (plus a veteran_flash_id association table).

Basically it boils down to this: `audit_event` table is tracking Claim History, and as such, the name change is unnecessary (unless you want it to semantically reflect that scope).
`claim` table is tracking current status.

This is my understanding. This is what we'll be going forward with.

**Comment by yoomlam** [comment]
(For my reference) [my feedback](https://dsva.slack.com/archives/C01Q7979Z7D/p1674237329439659?thread_ts=1674144499.310049&cid=C01Q7979Z7D) copied from Slack:

> Add requestId and idType to ClaimSubmission.

1. I think you mean: Add referenceId and idType to ClaimSubmission.
2. We should add the vbmsId on the ClaimEntity since MAS provides the benefitClaimId. This will be useful to look up the claim in VBMS and other VA systems. If claimId is meant to be the VBMS ID, please rename it to vbmsId.
3. offrampReason should be an explicit field (DB table column), not buried in details, so that reporting stats (using 3rd-party tools) is straightforward. offrampReason is mentioned in [Remove special issues to enable claim downstream routing](https://github.com/department-of-veterans-affairs/abd-vro/issues/431#top) #431 (and other tickets):
> When a claim is offramped, that offramp status and offrampReason should be saved to the DB.

4. presumptiveFlag should be an attribute of the ClaimEntity as stated in https://github.com/department-of-veterans-affairs/abd-vro/issues/446:
> ... the SaveAssessment step saves presumptive = true on the claim DB record ...

5. Add disabilityActionType to the ClaimEntity; also from 446:
> The DB should model/store disabilityActionType on the claim.

6. Add a list of flashIds to the VeteranEntity; also from 446:
> The DB should model the flashes associated with the Veteran. This is a list of integers.

7. Add icnTimestamp -- see [comment for 428](https://github.com/department-of-veterans-affairs/abd-vro/issues/428#issuecomment-1270578262):
> VRO should NOT be seen as the source-of-truth for the Veteran's ICN, so we should make that more explicit with this icnTimestamp DB field.

8. For [Exam-status endpoint for MAS/VBAAP to notify VRO](https://github.com/department-of-veterans-affairs/abd-vro/issues/434#top) #434 and [Exam ordering: VRO calls MAS/VBAAP to order an exam](https://github.com/department-of-veterans-affairs/abd-vro/issues/435#top) #435, you'll need to add an OrderExamEntity with field status
9.  For [VRO handles PDF upload to eFolder](https://github.com/department-of-veterans-affairs/abd-vro/issues/432#top) #432, add location info to EvidenceSummaryDocumentEntity:
> VRO saves to the DB the id, status, location, etc. (enough data for us to track down the PDF in eFolder) of the uploaded PDF.

10. For [Mark claim "RFD" to enable appropriate adjudicator assignment](https://github.com/department-of-veterans-affairs/abd-vro/issues/430#top) #430, consider adding an rfdFlag to ClaimSubmission. This can be used for reporting which and how many submitted claims were fast-tracked by VRO.

**Comment by yoomlam** [comment]
My responses:

> ## Open Questions:
> - `claim.id` is our internal DB primary key for the claim, but `claim.claim_submission_id` is an external number provided by [some external caller of our API]?

Yes. claimSubmissionId was to be renamed `referenceId` -- see my feedback item 1 above.
The `referenceId` and the `idType` together uniquely identify the claim within the external caller's system.

From [the wiki](https://github.com/department-of-veterans-affairs/abd-vro/wiki/VRO-Database#proposed-redesign):
> Add referenceId and idType to ClaimSubmission
> This is used to capture vendor-specific IDS such as collectionId for MAS

> - Is it better to think of `claim` table as "current claim state" vs `claim_submission` table as the new "claim audit event history"?

I was thinking of the `claim` table as representative of the claim as it is known in the system of record (VBMS) and other VA systems, which is why the `vbms_id`, `presumptive_flag`, and `disability_action_type` fields should exist in the `claim` table -- these values shouldn't change if the same claim was resubmitted. I think of the `claim_submission` table as the request's state/status as it is processed in VRO, where a new row is added for new requests even if it's for the same claim because VRO may fail on the first request (for whatever reason) but succeeds on the second request -- so the values for `off_ramp_reason` or `in_scope` may be different (b/c VRO logic changed).

>   - This might mean we can continue to store `details` as recorded history in `claim_submission`, but continue to maintain current state in `claim` table. Ultimately must consult with Yoom, but might be better to err on the side of redundancy than otherwise.

I don't (yet) see the value of storing the JSONB value if all the data can be stored as some DB field in the `claim` table or the `claim_submission` table.

>   - It feels like the root of this ambiguity is because there may be a misunderstanding on (or conflation with) claim history tracking vs current state

I believe the AuditEvent idea came from Rajesh. I wasn't made aware of its intent until I spoke with Dimitri 2 weeks ago. Afsin was also surprise that the existing tables created for VRO v1 was not being populated. The intent for v2 was to build on the DB tables from v1, not ignore them.

> - Do the data types look okay?
> - Where do we want any more indexing to happen?
> - What else am I missing?

I'll review the code to respond to these questions.

File: db-init/src/main/resources/database/migrations/V4.3__Update_Claim_Table.sql [hunk 1]
```diff
@@ -0,0 +1,8 @@
+ALTER TABLE claim 
+ADD COLUMN vbms_id VARCHAR, -- uuid?
+ADD COLUMN off_ramp_reason VARCHAR, 
+ADD COLUMN presumptive_flag BOOLEAN, 
+ADD COLUMN disability_action_type VARCHAR,
+ADD COLUMN in_scope BOOLEAN,
+ADD COLUMN submission_source VARCHAR,
```
Comment (yoomlam) [inline db-init/src/main/resources/database/migrations/V4.3__Update_Claim_Table.sql:7]: What is `submission_source`? I don't see it defined in the wiki.

File: db-init/src/main/resources/database/migrations/V4.3__Update_Claim_Table.sql [hunk 2]
```diff
@@ -0,0 +1,8 @@
+ALTER TABLE claim 
+ADD COLUMN vbms_id VARCHAR, -- uuid?
```
Comment (yoomlam) [inline db-init/src/main/resources/database/migrations/V4.3__Update_Claim_Table.sql:2]: Not a uuid

File: db-init/src/main/resources/database/migrations/V4.3__Update_Claim_Table.sql (hunk 1 above)
Comment (yoomlam) [reply db-init/src/main/resources/database/migrations/V4.3__Update_Claim_Table.sql:7]: All columns should have a comment to describe the field (bonus points if the comment provides sample values, particularly for VARCHAR that are enumerations in practice) -- https://www.postgresql.org/docs/current/sql-comment.html.

File: db-init/src/main/resources/database/migrations/V4.2__Update_AuditEvent_details_Column.sql [hunk 3]
```diff
@@ -0,0 +1,4 @@
+-- We do not need to preserve this data since we should not have any production audit details, yet
+ALTER TABLE audit_event DROP COLUMN details;
+-- As such, we can just recreate the column as JSONB
+ALTER TABLE audit_event ADD COLUMN details JSONB;
```
Comment (yoomlam) [inline db-init/src/main/resources/database/migrations/V4.2__Update_AuditEvent_details_Column.sql:4]: An AuditEvent (aka ClaimSubmission) should have a FK to a Contention (which is linked to the ClaimEntity relevant to the ClaimSubmission request) -- see diagram at https://github.com/department-of-veterans-affairs/abd-vro/wiki/VRO-Database#proposed-redesign

**Comment by alan-amida** [comment]
Just a note that I've had Ian make his changes in a separate work branch which will be pulled into this one:
#1060 

Going to summarize the new changes coming soon for clarity.

**Comment by alan-amida** [comment]
The overall plan for DB changes:
- [ ] KEEP THE `audit_event` table
  - [ ] add `reference_id` field
  - [ ] 
- [ ] Create `claim_submission` table
- [ ] Create `veteran_flash_id` table to stored a veteran's associated `flashId`s
  - currently only processing one flashId for Agent Orange exposure
- [ ] Create `exam_order` table to store the status of `/v2/examOrderingStatus` requests
- [ ] Update `audit_event` table's `details` column type from VARCHAR to JSONB
- [ ] Update our `claim` table with columned fields stored as JOSNB in `audit_event.details` (no replacement):
  - [ ] add `vbms_id` field
  - [ ] add `off_ramp_reason` field
  - [ ] add `presumptive_flag` field
  - [ ] add `disability_action_type` field
  - [ ] add `in_scope` field
  - ~~add `submission_source` field~~ (audit_event concern)
  - ~~add `submission_date` field~~ (audit_event concern)
- [ ] Update our `veteran` table to include a `icn_timestamp` field
- [ ] Update our `evidence_summary_document` table to include a `location` field

**Comment by alan-amida** [comment]
@yoomlam please review the above checklist to ensure we have everything correct.

**Comment by yoomlam** [comment]
> @yoomlam please review the above checklist to ensure we have everything correct.

That looks right.

**Comment by hayleymdawson** [comment]
One additional item from Slack as of 2/9/2023:

off_ramp_reason should be on the claim_submission table, and removed from the claim table.

**Comment by alan-amida** [comment]
Closing this PR in favor a new clean, commentless PR which can be found here: https://github.com/department-of-veterans-affairs/abd-vro/pull/1112

**Pull request closed without merging**